crates/red_knot/src/main.rs

```diff
@@ -48,225 +44,305 @@ fn main() -> anyhow::Result<()> {
-                            diagnostics.extend_from_slice(&file_diagnostics);
+impl MainLoop {
+    fn new() -> (Self, MainLoopCancellationToken) {
+        let (orchestrator_sender, orchestrator_receiver) = crossbeam_channel::bounded(1);
+        let (main_loop_sender, main_loop_receiver) = crossbeam_channel::bounded(1);
 
-                            if pending_analysis.fetch_sub(1, Ordering::SeqCst) == 1 {
-                                // Analysis completed, print the diagnostics.
-                                dbg!(&diagnostics);
-                            }
-                        }
+        let mut orchestrator = Orchestrator {
+            pending_analysis: None,
+            receiver: orchestrator_receiver,
+            sender: main_loop_sender.clone(),
+            aggregated_changes: AggregatedChanges::default(),
+        };
 
-                        Message::AnalysisCancelled => {
-                            if pending_analysis.fetch_sub(1, Ordering::SeqCst) == 1 {
-                                return consume_pending_messages(&receiver, aggregated_changes);
-                            }
-                        }
+        std::thread::spawn(move || {
+            orchestrator.run();
+        });
+
+        (
+            Self {
+                orchestrator_sender: orchestrator_sender.clone(),
+                main_loop_receiver,
+            },
+            MainLoopCancellationToken {
+                sender: main_loop_sender,
+            },
+        )
+    }
 
-                        Message::Exit => {
-                            run_cancellation_token_source.cancel();
+    fn file_changes_notifier(&self) -> FileChangesNotifier {
+        FileChangesNotifier {
+            sender: self.orchestrator_sender.clone(),
+        }
+    }
+
+    fn run(self, program: &mut Program) {
+        self.orchestrator_sender
+            .send(OrchestratorMessage::Run)
+            .unwrap();
+
+        for message in &self.main_loop_receiver {
+            tracing::trace!("Main Loop: Tick");
 
-                            // Don't consume any outstanding messages because we're exiting anyway.
-                            return NextTickCommand::Exit;
+            match message {
+                MainLoopMessage::CheckProgram => {
+                    // Remove mutability from program.
+                    let program = &*program;
+                    let run_cancellation_token_source = CancellationTokenSource::new();
+                    let run_cancellation_token = run_cancellation_token_source.token();
+                    let sender = &self.orchestrator_sender;
+
+                    sender
+                        .send(OrchestratorMessage::CheckProgramStarted {
+                            cancellation_token: run_cancellation_token_source,
+                        })
+                        .unwrap();
+
+                    rayon::in_place_scope(|scope| {
```

Comment:
Yes, that's correct. Although the `ApplyChanges` might gets debounced first to wait for new incoming file change message. It would be nice if this could all be in a single place but the challenge I faced is that we need to have a single "input" on which the orchestrator thread can wait on (suspend).